### PR TITLE
Add CLI flags to optionally connect to an external etcd.

### PR DESCRIPTION
**Pull request by dgoodwin**
Flags and TLS setup based on code from kube-apiserver.

Tested and should work with both TLS enabled etcd and plain http.

### CODE REVIEW

File: cmd/kcp/kcp.go [hunk 1]
```diff
@@ -219,7 +223,37 @@ func main() {
...
+					etcdClientInfo.TLS.Certificates = []tls.Certificate{cert}
+				}
+
+				if len(etcdClientInfo.TrustedCAFile) > 0 {
+					if caCert, err := ioutil.ReadFile(etcdClientInfo.TrustedCAFile); err != nil {
+						return fmt.Errorf("failed to read ca file: %s", err)
+					} else {
+						caPool := x509.NewCertPool()
+						caPool.AppendCertsFromPEM(caCert)
+						etcdClientInfo.TLS.RootCAs = caPool
```
Comment (pires) [inline cmd/kcp/kcp.go:250]: At this point, wouldn't you want to set `etcdClientInfo.TLS.InsecureSkipVerify = false`?

File: cmd/kcp/kcp.go [hunk 2]
```diff
@@ -229,8 +263,17 @@ func main() {
...
-	cmd.AddCommand(startCmd)
 
+	startCmd.Flags().StringSliceVar(&etcdClientInfo.Endpoints, "etcd-servers", etcdClientInfo.Endpoints,
+		"List of external etcd servers to connect with (scheme://ip:port), comma separated. If absent an in-process etcd will be created.")
+	startCmd.Flags().StringVar(&etcdClientInfo.KeyFile, "etcd-keyfile", etcdClientInfo.KeyFile,
+		"SSL key file used to secure etcd communication.")
+	startCmd.Flags().StringVar(&etcdClientInfo.CertFile, "etcd-certfile", etcdClientInfo.CertFile,
+		"SSL certification file used to secure etcd communication.")
+	startCmd.Flags().StringVar(&etcdClientInfo.TrustedCAFile, "etcd-cafile", etcdClientInfo.TrustedCAFile,
+		"SSL Certificate Authority file used to secure etcd communication.")
```
Comment (pires) [inline cmd/kcp/kcp.go:274]: Can we take this opportunity to replace _SSL_ with _TLS_? 👮🏻‍♂️

**Review by pires — COMMENTED** [review]
Would be cool to actually implement e2e test that brings up an out-of-process etcd _a la_ [kubebuilder and envtest](https://book.kubebuilder.io/reference/envtest.html)?

File: cmd/kcp/kcp.go (hunk 1 above)
Comment (dgoodwin) [reply cmd/kcp/kcp.go:250]: Good catch, I had that in there at one point and deleted it somehow. Thanks!

File: cmd/kcp/kcp.go (hunk 2 above)
Comment (dgoodwin) [reply cmd/kcp/kcp.go:274]: SGTM.

**Review by davidfestal — APPROVED** [review]
That's nice. Thanks !

File: cmd/kcp/kcp.go [hunk 3]
```diff
@@ -219,7 +223,38 @@ func main() {
 				prepared := server.PrepareRun()
 
 				return prepared.Run(ctx.Done())
-			})
+			}
+
+			if len(etcdClientInfo.Endpoints) == 0 {
+				// No etcd servers specified so create one in-process:
+				return s.Run(runFunc)
+			} else {
```
Comment (imjasonh) [inline cmd/kcp/kcp.go:231]: style nit: Since the previous `if` returns, you can skip the `else` here and unindent the rest of this.

**Review by imjasonh — COMMENTED** [review]
Nice!

**Comment by dgoodwin** [comment]
Updated for the style change, thanks!

**Review by imjasonh — APPROVED** [review]

**Comment by imjasonh** [comment]
Thanks for adding this! 👍

**Pull request merged**